proto/substrait/algebra.proto

```diff
@@ -605,6 +606,31 @@ message MergeJoinRel {
   substrait.extensions.AdvancedExtension advanced_extension = 10;
 }
 
+// The nested loop join (NLJ) operator will hold the entire right input and iterate over it using the
+// left input, evaluating the join expression on the Cartesian product of all rows.
+message NestedLoopJoinRel {
+  RelCommon common = 1;
+  Rel left = 2;
+  Rel right = 3;
+  Expression expression = 4;
+
+  JoinType type = 5;
+
+  enum JoinType {
+    JOIN_TYPE_UNSPECIFIED = 0;
+    JOIN_TYPE_INNER = 1;
+    JOIN_TYPE_OUTER = 2;
+    JOIN_TYPE_LEFT = 3;
+    JOIN_TYPE_RIGHT = 4;
+    JOIN_TYPE_LEFT_SEMI = 5;
+    JOIN_TYPE_RIGHT_SEMI = 6;
+    JOIN_TYPE_LEFT_ANTI = 7;
+    JOIN_TYPE_RIGHT_ANTI = 8;
```

Comment:
I [historically](https://github.com/substrait-io/substrait/pull/467) don't like anti joins because they are not well defined.

For example, a semi join can be implemented as:

```
for l_row in left:
  for r_row in right:
    result = expression(l_row, r_row):
    if result == True:
      yield l_row
```

However, when we look at the anti join:

```
for l_row in left:
  for r_row in right:
    result = expression(l_row, r_row)
    if ???:
      yield l_row
```

Should `???` be `result == False` or should it be `result != True`?  Specifically the question is around how we handle the case where `result` is `null`.

Do you want to take a stab at either picking one of these and clearly documenting this or adding a new proto field allowing the user to choose between these two behaviors?

That being said, I won't hold this issue up over this, so if you'd rather punt on the question then that is fine too.

Reply:
Very illustrative point, thank you. I am in favor of allowing the user to define NULL behavior for the anti-join. However, I would prefer to push this to a separate PR where JoinType is refactored out for all of NLJ/Hash/SM operators (I assume this should be applied to all three). I can create a separate issue for that if you think it's appropriate?

Reply:
I created https://github.com/substrait-io/substrait/issues/563